main/lsp/ExtractVariable.cc

```diff
@@ -0,0 +1,239 @@
+        auto &rescue = ast::cast_tree_nonnull<ast::Rescue>(tree);
+        updateEnclosingScope(tree, rescue.body.loc());
+        updateEnclosingScope(tree, rescue.else_.loc());
+        updateEnclosingScope(tree, rescue.ensure.loc());
+    }
+
+    void preTransformRescueCase(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &rescueCase = ast::cast_tree_nonnull<ast::RescueCase>(tree);
+        updateEnclosingScope(tree, rescueCase.body.loc());
+        skipLoc(rescueCase.var.loc());
+        // TODO: skipLoc for exceptions
+    }
+
+    void preTransformWhile(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &while_ = ast::cast_tree_nonnull<ast::While>(tree);
+        updateEnclosingScope(tree, while_.body.loc());
+    }
+};
+
+vector<unique_ptr<TextDocumentEdit>> getExtractVariableEdits(LSPTypecheckerDelegate &typechecker,
+                                                             const LSPConfiguration &config,
+                                                             unique_ptr<Range> selectionRange,
+                                                             const core::Loc selectionLoc) {
+    auto loc = selectionLoc;
+    auto file = loc.file();
+    const core::GlobalState &gs = typechecker.state();
+    vector<unique_ptr<TextEdit>> edits;
+
+    ExtractVariableWalk extractVariableWalk(loc);
+    auto desugaredTree = typechecker.getDesugared(file);
+    core::Context ctx(gs, core::Symbols::root(), file);
+    ast::TreeWalk::apply(ctx, extractVariableWalk, desugaredTree);
+
+    if (extractVariableWalk.foundExactMatch()) {
+        vector<unique_ptr<TextEdit>> edits;
+
+        auto locOffsets = loc.offsets();
+        core::LocOffsets whereToInsert = core::LocOffsets::none();
+        auto enclosingScope = extractVariableWalk.enclosingScope;
+        // For all cases except InsSeq and ClassDef, extractVariableWalk.enclosingScopeLoc should be
+        // the same as what we're pulling out from the ExpressionPtr, but we'll just get it directly
+        // for consistency.
+        if (auto insSeq = ast::cast_tree<ast::InsSeq>(*enclosingScope)) {
+            for (auto &stat : insSeq->stats) {
+                if (stat.loc().contains(locOffsets)) {
+                    ENFORCE(!ast::isa_tree<ast::InsSeq>(stat));
+                    whereToInsert = stat.loc();
+                    break;
+                }
+            }
+            if (insSeq->expr.loc().contains(locOffsets)) {
+                ENFORCE(!ast::isa_tree<ast::InsSeq>(insSeq->expr));
+                whereToInsert = insSeq->expr.loc();
+            }
+            ENFORCE(whereToInsert.exists());
+        } else if (auto block = ast::cast_tree<ast::Block>(*enclosingScope)) {
+            ENFORCE(!ast::isa_tree<ast::InsSeq>(block->body));
+            whereToInsert = block->body.loc();
+        } else if (auto methodDef = ast::cast_tree<ast::MethodDef>(*enclosingScope)) {
+            // TODO(neil): this will fail for endless methods
```

Comment:
Endless methods aren't supported in pay-server yet, so we can handle this later.